internal/provider/services/storage/storage.go

```diff
@@ -0,0 +1,235 @@
+	DefaultChunkSize                     int64 = 64 * 1024 // 64 KB
+)
+
+var (
+	BucketName             = DefaultBucketName
+	StorageServiceEndpoint = DefaultStorageServiceDefaultEndpoint
+	ChunkSize              = DefaultChunkSize
+
+	BucketInitializationError = status.Error(codes.Unknown, "Failed to initialize bucket for storage.")
+	DownloadSourceZipError    = status.Error(codes.Unknown, "Failed to download source zip.")
+	SendSourceZipError        = status.Error(codes.Unknown, "Failed to send source zip.")
+	SendShasumError           = status.Error(codes.Unknown, "Failed to send shasum file.")
+	ContentLengthError        = status.Error(codes.Unknown, "Failed to read correct content length.")
+	DownloadShasumError       = status.Error(codes.Unknown, "Failed to download shasum.")
+)
+
+type StorageService struct {
+	services.UnimplementedStorageServer
+	Client     storage.AWSS3BucketClient
+	BucketName string
+	Region     string
+}
+
+// Registers StorageService with grpc server
+func (s *StorageService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) error {
+	if err := storage.InitializeS3Bucket(s.BucketName, s.Region, s.Client); err != nil {
+		log.Println(err)
+		return BucketInitializationError
+	}
+
+	services.RegisterStorageServer(grpcServer, s)
+
+	return nil
+}
+
+// Download Source Zip from storage
+func (s *StorageService) DownloadProviderSourceZip(request *services.DownloadSourceZipRequest, server services.Storage_DownloadProviderSourceZipServer) error {
+
+	log.Println("Downloading source zip.")
+
+	ctx := server.Context()
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("provider.name", request.GetProvider().GetName()),
+		attribute.String("provider.version", request.GetProvider().GetVersion()),
+		attribute.String("provider.os", request.GetProvider().GetOs()),
+		attribute.String("provider.arch", request.GetProvider().GetArch()),
+	)
+
+	providerAddress := strings.Split(request.Provider.GetName(), "/")
+	filename := fmt.Sprintf("terraform-provider-%s_%s_%s_%s.zip", providerAddress[1], request.GetProvider().GetVersion(), request.GetProvider().GetOs(), request.GetProvider().GetArch())
+	fileLocation := services.ResolveS3Locations(request.Provider.GetName(), request.Provider.GetVersion(), filename)
+
+	in := &s3.GetObjectInput{
+		Bucket: aws.String(BucketName),
+		Key:    aws.String(fileLocation),
+	}
+
+	out, err := s.Client.GetObject(ctx, in)
+	if err != nil {
+		span.RecordError(err)
+		log.Println(err)
+		return DownloadSourceZipError
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(out.Body)
+	bb := buf.Bytes()
+
+	outContentLength := out.ContentLength
+
+	if int64(len(bb)) == outContentLength {
+		res := &services.SourceZipResponse{}
+		for i := int64(0); i < outContentLength; i += ChunkSize {
+			if i+ChunkSize > outContentLength {
+				res.ZipDataChunk = bb[i:outContentLength]
+			} else {
+				res.ZipDataChunk = bb[i : i+ChunkSize]
+			}
+
+			if err := server.Send(res); err != nil {
```

Comment:
By reading the whole file into memory in one go and then send chunks you're using more memory than you need to - these provider zip files can be big files!
Better to read a chunk from the out.Body send that, then repeat.

Reply:
hmm, modified the chunking logic of the code to read a chunk and send it repetatively until the end of the file is reached.